### ISSUE TITLE
Create Source and Drain Objects

### ISSUE THREAD

**Issue by Victor2266** [Victor2266/CPU-GPU-Fluid-Simulation-In-Unity#16]
Drain Object: Fluids which touch this should deactivate
Source Object: Deactivated particles should have their position set to this object and become activated again, their velocity should point away from the source, they should spawn up to the maximum allowed particles in the scene. This gives the illusion of particles spawning in from the source.

**Comment by jj-marr** [comment]
Can we have more than one source/drain in a scene? How would that work? Can we create a queue/off-screen buffer of particles that the drain empties into?

**Comment by Victor2266** [comment]
- Yes we should be able to have more than one source/drain in a scene
    - The drain object will just deactivate the particles which touch it via the fluidType field that each particle struct has, those particles will be 'skipped over' in the simulation and not displayed on the screen
    - The source object iterates through the list of particles, if the particle is active already then it will skip it, if a particle is inactive then it will activate it and place it in the source output position

Let me know if there is a more efficient way.

Keeping track of the inactive particles would speed up the traversal of the source object, but due to the GPU memory allocation you would need to initialize a buffer which can hold the max possible quantity of inactive particles which is equal to the max number of particles in total. so I'm not sure which is a better approach.

**Comment by jj-marr** [comment]
Can the source output more than one particle per simulation step?

**Comment by Victor2266** [comment]
Yes you could theoretically make it create anywhere from 1 to the maximum number of particles / per simulation Step

**Comment by Victor2266** [comment]
This task is blocked by needing to implement active and inactive particles

**Comment by Victor2266** [comment]
i completed #55 which can be used as a guide for the drain objects

**Comment by Victor2266** [comment]
#43 is done so this is unblocked

**Comment by Victor2266** [comment]
Drain objects will be oriented box colliders, because you probably want to section areas for draining
Source objects will be circle colliders, because it is easier to spawn particle around a single point

**Comment by Victor2266** [comment]
I'm marking this as done because they work right now, but the source objects are hard coded to spawn in only water, I want to look into ECS more before I figure out a way to set fluid types per source object
